diagnostic/conflict.go

```diff
@@ -72,6 +77,36 @@ func groupConflicts(allConflicts []conflict) []conflict {
 			key = p.producerRepr + ";" + p.consumerRepr
 			if p.producerPosition.IsValid() {
 				key = p.producerPosition.String() + ": " + p.producerRepr
+			} else {
+				// Find the enclosing function name for the conflict position. Prepend that to the default key for a tighter
+				// grouping condition.
+				conf := pass.ResultOf[config.Analyzer].(*config.Config)
+				for _, file := range pass.Files {
+					// `fileName` stores the complete file path relative to the current working directory
+					fileName := pass.Fset.Position(file.FileStart).Filename
+					if fn, err := filepath.Rel(cwd, fileName); err == nil {
+						fileName = fn
+					}
+					// Check if the file is in scope and the conflict position is in the same file
+					if !conf.IsFileInScope(file) || fileName != c.position.Filename {
+						continue
+					}
+
+					// Find the enclosing function name for the conflict position
+					ast.Inspect(file, func(n ast.Node) bool {
+						if fd, ok := n.(*ast.FuncDecl); ok {
+							// Check if the conflict position falls within the function's position range. If so, update
+							// the key to include the function name, and end the traversal.
+							functionStart := pass.Fset.Position(fd.Pos()).Offset
+							functionEnd := pass.Fset.Position(fd.End()).Offset
+							if c.position.Offset >= functionStart && c.position.Offset <= functionEnd {
+								key = fd.Name.Name + ":" + key
+								return false
+							}
+						}
+						return true
+					})
+				}
```

Comment:
I think this AST search can be simplified to just a single pass over `ast.File` node since function declarations are top-level decls and can only appear under `ast.File` node: https://go.dev/ref/spec#Declarations_and_scope

It would save some further traversals :)

Reply:
Great idea! Updated the code accordingly.